[BE] Correctly catch skip signals emitting from sys.exit

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129581

Some tests in test_c10d_nccl.py overwrite `_join_process()` and `_check_return_codes()`, which cause the skip signals are not catched appropriately. This PR fixes the issue.

Differential Revision: [D59067457](https://our.internmc.facebook.com/intern/diff/D59067457/)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

fegin committed:
[BE] Correctly catch skip signals emitting from sys.exit

Some tests in test_c10d_nccl.py overwrite `_join_process()` and `_check_return_codes()`, which cause the skip signals are not catched appropriately. This PR fixes the issue.

Differential Revision: [D59067457](https://our.internmc.facebook.com/intern/diff/D59067457/)

[ghstack-poisoned]

fegin committed:
Update on "[BE] Correctly catch skip signals emitting from sys.exit"


Some tests in test_c10d_nccl.py overwrite `_join_process()` and `_check_return_codes()`, which cause the skip signals are not catched appropriately. This PR fixes the issue.

Differential Revision: [D59067457](https://our.internmc.facebook.com/intern/diff/D59067457/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -4014,6 +4018,10 @@ def _create_process_group_nccl(self):
         return pg
 
     def _check_return_codes(self, elapsed_time):
+        for skip in TEST_SKIPS.values():
+            if self.processes[0].exitcode == skip.exit_code:
+                return MultiProcessTestCase()._check_return_codes(self)
+
```
We have a lot of discussions around this. Basically for the test skipped, are we ok that the parent process does not know it and let the child process directly skip it? (Maybe, because we don't really care whether the test is skipped or not because when we don't have the matched environment, the test cannot be run anyway. Any added value there? Probably not). But this somehow break the original design which aims at letting parent to know everything.

For now, to unblock the fix, maybe let's make this a decorator or util function so that we know down the road, when we override  `_check_return_codes` we need to handle skip case as well.

fduwjj reviewed (approved):
Stamp for now and please don't copy paste same code because we might have more down the road. Thanks!

fegin committed:
Update on "[BE] Correctly catch skip signals emitting from sys.exit"


Some tests in test_c10d_nccl.py overwrite `_join_process()` and `_check_return_codes()`, which cause the skip signals are not catched appropriately. This PR fixes the issue.

Differential Revision: [D59067457](https://our.internmc.facebook.com/intern/diff/D59067457/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

fegin committed:
Update on "[BE] Correctly catch skip signals emitting from sys.exit"


Some tests in test_c10d_nccl.py overwrite `_join_process()` and `_check_return_codes()`, which cause the skip signals are not catched appropriately. This PR fixes the issue.

Differential Revision: [D59067457](https://our.internmc.facebook.com/intern/diff/D59067457/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

fegin committed:
Update on "[BE] Correctly catch skip signals emitting from sys.exit"


Some tests in test_c10d_nccl.py overwrite `_join_process()` and `_check_return_codes()`, which cause the skip signals are not catched appropriately. This PR fixes the issue.

Differential Revision: [D59067457](https://our.internmc.facebook.com/intern/diff/D59067457/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

fegin committed:
Update on "[BE] Correctly catch skip signals emitting from sys.exit"


Some tests in test_c10d_nccl.py overwrite `_join_process()` and `_check_return_codes()`, which cause the skip signals are not catched appropriately. This PR fixes the issue.

Differential Revision: [D59067457](https://our.internmc.facebook.com/intern/diff/D59067457/)

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]
